frontend/public/components/modals/taint-update-modal.tsx

```diff
@@ -0,0 +1,116 @@
+/* eslint-disable no-undef, no-unused-vars */
+
+import * as React from 'react';
+import { PromiseComponent } from '../utils';
+import { K8sKind, k8sPatch } from '../../module/k8s';
+
+import {createModalLauncher, ModalBody, ModalComponentProps, ModalSubmitFooter, ModalTitle} from '../factory';
+import * as _ from 'lodash-es';
+
+class TaintUpdateModal extends PromiseComponent {
+  readonly state: TaintUpdateModalState;
+
+  constructor(public props: TaintUpdateModalProps) {
+    super(props);
+    this.state.taints = this.props.resource.spec.taints || [];
+  }
+
+  _submit = (e) => {
+    e.preventDefault();
+
+    // We just throw away any rows where the key is blank
+    const taints = _.reject(this.state.taints, t => _.isEmpty(t.key));
+
+    const keys = taints.map(t => t.key);
+    if (_.uniq(keys).length !== keys.length) {
+      this.setState({errorMessage: 'Duplicate keys found.'});
+      return;
+    }
+
+    // Convert any blank values to null
+    _.each(taints, t => t.value = _.isEmpty(t.value) ? null : t.value);
+
+    // Make sure to 'add' if the path does not already exist, otherwise the patch request will fail
+    const op = this.props.resource.spec.taints ? 'replace' : 'add';
+    const patch = [{path: '/spec/taints', op, value: taints}];
+
+    this.handlePromise(
+      k8sPatch(this.props.resourceKind, this.props.resource, patch)
+    ).then(this.props.close, () => {});
+  };
+
+  _cancel = () => {
+    this.props.close();
+  };
+
+  _change = (e, i, field) => {
+    const clonedTaints = _.cloneDeep(this.state.taints);
+    clonedTaints[i][field] = e.target ? e.target.value : e;
+    this.setState({taints: clonedTaints});
+  };
+
+  _remove = (i) => {
+    const clonedTaints = _.cloneDeep(this.state.taints);
+    clonedTaints.splice(i, 1);
+    this.setState({taints: clonedTaints});
+  };
+
+  _addRow = () => {
+    const clonedTaints = _.cloneDeep(this.state.taints);
+    clonedTaints.push({key: '', value: '', effect: ''});
+    this.setState({taints: clonedTaints});
```

Comment:
This should use a `setState` callback. It's needed when updating state based on the previous value since `this.state` can be stale.

```ts
  this.setState((state: TaintUpdateModalState) => ({
    taints: [...state.taints, {key: '', value: '', effect: ''}],
  }));
```

https://reactjs.org/docs/react-component.html#setstate